Node hangs during start up in test_total_space_limit_of_commitlog_with_medium_limit

Seen first in https://jenkins.scylladb.com/view/master/job/scylla-master/job/dtest-daily-release/71/artifact/logs-full.release.001/1659755830552_commitlog_test.py%3A%3ATestCommitLog%3A%3Atest_total_space_limit_of_commitlog_with_medium_limit/node1.log
```
Scylla version 5.2.0~dev-0.20220805.fbbe2529c11d with build-id be8ad53ed567b044861b8b8fa8042e72d818c935 starting ...

INFO  2022-08-06 03:05:59,631 [shard 0] init - initializing batchlog manager
INFO  2022-08-06 03:05:59,632 [shard 0] system_keyspace - Loaded local host id: b30f75cb-725a-42b6-a031-eb8b5071ccbd
INFO  2022-08-06 03:05:59,633 [shard 0] format_selector - Selected me sstables format
INFO  2022-08-06 03:15:02,419 [shard 0] compaction_manager - Asked to stop
INFO  2022-08-06 03:15:02,419 [shard 0] compaction_manager - Stopped
INFO  2022-08-06 03:15:02,419 [shard 1] compaction_manager - Asked to stop
INFO  2022-08-06 03:15:02,419 [shard 1] compaction_manager - Stopped
```

Note the ~9 minutes gap between the last start up message and the first shutdown message
when the dtest tries to kill the node when the test fails on timeout.
Shutting down the node times out too and a core dump is produced: https://jenkins.scylladb.com/view/master/job/scylla-master/job/dtest-daily-release/71/artifact/logs-full.release.001/1659755830552_commitlog_test.py%3A%3ATestCommitLog%3A%3Atest_total_space_limit_of_commitlog_with_medium_limit/node1-scylla.32900.1659755829.core.gz

This seems to be highly reproducible, at least on the Jenkins runs.

For completeness, here's the corresponding dtest log from
https://jenkins.scylladb.com/view/master/job/scylla-master/job/dtest-daily-release/71/artifact/logs-full.release.001/dtest-gw1.log
```
03:05:39,670 390     ccm                            DEBUG    cluster.py          :704  | test_total_space_limit_of_commitlog_with_medium_limit: node1: Starting scylla: args=['/jenkins/workspace/scylla-master/dtest-daily-release/scylla/.dtest/dtest-kulhml98/test/node1/bin/scylla', '--options-file', '/jenkins/workspace/scylla-master/dtest-daily-release/scylla/.dtest/dtest-kulhml98/test/node1/conf/scylla.yaml', '--log-to-stdout', '1', '--smp', '1', '--api-address', '127.0.57.1', '--collectd-hostname', '9d540a57aee7.node1', '--memory', '1024M', '--abort-on-seastar-bad-alloc', '--abort-on-lsa-bad-alloc', '1', '--abort-on-internal-error', '1', '--developer-mode', 'true', '--default-log-level', 'info', '--collectd', '0', '--overprovisioned', '--prometheus-address', '127.0.57.1', '--unsafe-bypass-fsync', '1', '--kernel-page-cache', '1', '--max-networking-io-control-blocks', '1000'] wait_other_notice=True wait_for_binary_proto=True
03:05:40,530 390     ccm                            DEBUG    cluster.py          :704  | test_total_space_limit_of_commitlog_with_medium_limit: node1: Starting scylla-jmx: args=['/jenkins/workspace/scylla-master/dtest-daily-release/scylla/.dtest/dtest-kulhml98/test/node1/bin/symlinks/scylla-jmx', '-Dapiaddress=127.0.57.1', '-Djavax.management.builder.initial=com.scylladb.jmx.utils.APIBuilder', '-Djava.rmi.server.hostname=127.0.57.1', '-Dcom.sun.management.jmxremote', '-Dcom.sun.management.jmxremote.host=127.0.57.1', '-Dcom.sun.management.jmxremote.port=7199', '-Dcom.sun.management.jmxremote.rmi.port=7199', '-Dcom.sun.management.jmxremote.local.only=false', '-Xmx256m', '-XX:+UseSerialGC', '-Dcom.sun.management.jmxremote.authenticate=false', '-Dcom.sun.management.jmxremote.ssl=false', '-jar', '/jenkins/workspace/scylla-master/dtest-daily-release/scylla/.dtest/dtest-kulhml98/test/node1/bin/scylla-jmx-1.0.jar']
03:05:40,896 390     commitlog_test                 DEBUG    commitlog_test.py   :877  | test_total_space_limit_of_commitlog_with_medium_limit: Create test keyspace and table
03:05:40,897 390     cassandra.cluster              WARNING  dtest_setup.py      :386  | test_total_space_limit_of_commitlog_with_medium_limit: Cluster.__init__ called with contact_points specified, but load-balancing policies are not specified in some ExecutionProfiles. In the next major version, this will raise an error; please specify a load-balancing policy. (contact_points = ['127.0.57.1'], EPs without explicit LBPs = ('EXEC_PROFILE_DEFAULT',))
03:05:40,903 390     cassandra.policies             INFO     policies.py         :289  | test_total_space_limit_of_commitlog_with_medium_limit: Using datacenter 'datacenter1' for DCAwareRoundRobinPolicy (via host '127.0.57.1:9042'); if incorrect, please specify a local_dc to the constructor, or limit contact points to local cluster nodes
03:05:40,929 390     commitlog_test                 DEBUG    commitlog_test.py   :885  | test_total_space_limit_of_commitlog_with_medium_limit: Starting CS workload: ['write', 'n=204800', '-rate', 'threads=25', '-col', 'size=FIXED(1000)', '-log', 'interval=100']
03:05:46,155 390     commitlog_test                 DEBUG    commitlog_test.py   :914  | test_total_space_limit_of_commitlog_with_medium_limit: Current commitlog size: 600M (58.59)%, Well-used cases: 1 / 5, Time passed: 5.23s, commitlog_disk_usage_threshold: 512M
03:05:57,716 390     commitlog_test                 DEBUG    commitlog_test.py   :924  | test_total_space_limit_of_commitlog_with_medium_limit: ('******************** Stress Settings ********************\nCommand:\n  Type: write\n  Count: 204,800\n  No Warmup: false\n  Consistency Level: LOCAL_ONE\n  Serial Consistency Level: SERIAL\n  Target Uncertainty: not applicable\n  Key Size (bytes): 10\n  Counter Increment Distibution: add=fixed(1)\nRate:\n  Auto: false\n  Thread Count: 25\n  OpsPer Sec: 0\nPopulation:\n  Sequence: 1..204800\n  Order: ARBITRARY\n  Wrap: true\nInsert:\n  Revisits: Uniform:  min=1,max=1000000\n  Visits: Fixed:  key=1\n  Row Population Ratio: Ratio: divisor=1.000000;delegate=Fixed:  key=1\n  Batch Type: not batching\nColumns:\n  Max Columns Per Key: 5\n  Column Names: [C0, C1, C2, C3, C4]\n  Comparator: AsciiType\n  Timestamp: null\n  Variable Column Count: false\n  Slice: false\n  Size Distribution: Fixed:  key=1000\n  Count Distribution: Fixed:  key=5\nErrors:\n  Ignore: false\n  Tries: 10\nLog:\n  No Summary: false\n  No Settings: false\n  File: null\n  Interval Millis: 100000\n  Level: NORMAL\nMode:\n  API: JAVA_DRIVER_NATIVE\n  Connection Style: CQL_PREPARED\n  CQL Version: CQL3\n  Protocol Version: V4\n  Username: null\n  Password: null\n  Auth Provide Class: null\n  Max Pending Per Connection: null\n  Connections Per Host: 8\n  Compression: NONE\nNode:\n  Nodes: [127.0.57.1]\n  Is White List: false\n  Datacenter: null\nSchema:\n  Keyspace: keyspace1\n  Replication Strategy: org.apache.cassandra.locator.SimpleStrategy\n  Replication Strategy Options: {replication_factor=1}\n  Table Compression: null\n  Table Compaction Strategy: null\n  Table Compaction Strategy Options: {}\nTransport:\n  factory=org.apache.cassandra.thrift.TFramedTransportFactory; truststore=null; truststore-password=null; keystore=null; keystore-password=null; ssl-protocol=TLS; ssl-alg=SunX509; store-type=JKS; ssl-ciphers=TLS_RSA_WITH_AES_128_CBC_SHA,TLS_RSA_WITH_AES_256_CBC_SHA; \nPort:\n  Native Port: 9042\n  Thrift Port: 9160\n  JMX Port: 7199\nSend To Daemon:\n  *not set*\nGraph:\n  File: null\n  Revision: unknown\n  Title: null\n  Operation: WRITE\nTokenRange:\n  Wrap: false\n  Split Factor: 1\n\n===== Using optimized driver!!! =====\nConnected to cluster: test, max pending requests per connection null, max connections per host 8\nDatatacenter: datacenter1; Host: /127.0.57.1; Rack: rack1\nCreated keyspaces. Sleeping 1s for propagation.\nSleeping 2s...\nWarming up WRITE with 50000 iterations...\nRunning WRITE with 25 threads for 204800 iteration\ntype       total ops,    op/s,    pk/s,   row/s,    mean,     med,     .95,     .99,    .999,     max,   time,   stderr, errors,  gc: #,  max ms,  sum ms,  sdv ms,      mb\ntotal,        204800,   23814,   23814,   23814,     1.0,     1.0,     2.0,     2.1,     4.0,    11.6,    8.6,  0.00000,      0,      0,       0,       0,       0,       0\n\n\nResults:\nOp rate                   :   23,814 op/s  [WRITE: 23,814 op/s]\nPartition rate            :   23,814 pk/s  [WRITE: 23,814 pk/s]\nRow rate                  :   23,814 row/s [WRITE: 23,814 row/s]\nLatency mean              :    1.0 ms [WRITE: 1.0 ms]\nLatency median            :    1.0 ms [WRITE: 1.0 ms]\nLatency 95th percentile   :    2.0 ms [WRITE: 2.0 ms]\nLatency 99th percentile   :    2.1 ms [WRITE: 2.1 ms]\nLatency 99.9th percentile :    4.0 ms [WRITE: 4.0 ms]\nLatency max               :   11.6 ms [WRITE: 11.6 ms]\nTotal partitions          :    204,800 [WRITE: 204,800]\nTotal errors              :          0 [WRITE: 0]\nTotal GC count            : 0\nTotal GC memory           : 0.000 KiB\nTotal GC time             :    0.0 seconds\nAvg GC time               :    NaN ms\nStdDev GC time            :    0.0 ms\nTotal operation time      : 00:00:08\n\nEND\n', '')
03:05:58,731 390     commitlog_test                 DEBUG    commitlog_test.py   :927  | test_total_space_limit_of_commitlog_with_medium_limit: Final commitlog size: [/jenkins/workspace/scylla-master/dtest-daily-release/scylla/.dtest/dtest-kulhml98/test/node1/commitlogs] 1001M
03:05:58,774 390     commitlog_test                 DEBUG    commitlog_test.py   :934  | test_total_space_limit_of_commitlog_with_medium_limit: Restart node1 to enable default commitlog configure
03:05:58,780 390     ccm                            DEBUG    cluster.py          :704  | test_total_space_limit_of_commitlog_with_medium_limit: node1: Starting scylla: args=['/jenkins/workspace/scylla-master/dtest-daily-release/scylla/.dtest/dtest-kulhml98/test/node1/bin/scylla', '--options-file', '/jenkins/workspace/scylla-master/dtest-daily-release/scylla/.dtest/dtest-kulhml98/test/node1/conf/scylla.yaml', '--log-to-stdout', '1', '--api-address', '127.0.57.1', '--collectd-hostname', '9d540a57aee7.node1', '--smp', '2', '--memory', '1024M', '--abort-on-seastar-bad-alloc', '--abort-on-lsa-bad-alloc', '1', '--abort-on-internal-error', '1', '--developer-mode', 'true', '--default-log-level', 'info', '--collectd', '0', '--overprovisioned', '--prometheus-address', '127.0.57.1', '--unsafe-bypass-fsync', '1', '--kernel-page-cache', '1', '--max-networking-io-control-blocks', '1000'] wait_other_notice=True wait_for_binary_proto=True
03:06:10,913 390     cassandra.cluster              WARNING  thread.py           :58   | test_total_space_limit_of_commitlog_with_medium_limit: Host 127.0.57.1:9042 has been marked down
03:06:10,914 390     cassandra.cluster              WARNING  thread.py           :58   | test_total_space_limit_of_commitlog_with_medium_limit: [control connection] Error connecting to 127.0.57.1:9042:
Traceback (most recent call last):
  File "cassandra/cluster.py", line 3504, in cassandra.cluster.ControlConnection._reconnect_internal
  File "cassandra/cluster.py", line 3526, in cassandra.cluster.ControlConnection._try_connect
  File "cassandra/cluster.py", line 1613, in cassandra.cluster.Cluster.connection_factory
  File "cassandra/connection.py", line 758, in cassandra.connection.Connection.factory
  File "/usr/local/lib64/python3.9/site-packages/cassandra/io/libevreactor.py", line 267, in __init__
    self._connect_socket()
  File "cassandra/connection.py", line 825, in cassandra.connection.Connection._connect_socket
ConnectionRefusedError: [Errno 111] Tried connecting to [('127.0.57.1', 9042)]. Last error: Connection refused
03:06:11,916 390     cassandra.pool                 WARNING  thread.py           :58   | test_total_space_limit_of_commitlog_with_medium_limit: Error attempting to reconnect to 127.0.57.1:9042, scheduling retry in 2.24 seconds: [Errno 111] Tried connecting to [('127.0.57.1', 9042)]. Last error: Connection refused
03:06:14,219 390     cassandra.pool                 WARNING  thread.py           :58   | test_total_space_limit_of_commitlog_with_medium_limit: Error attempting to reconnect to 127.0.57.1:9042, scheduling retry in 3.88 seconds: [Errno 111] Tried connecting to [('127.0.57.1', 9042)]. Last error: Connection refused
03:06:18,124 390     cassandra.pool                 WARNING  thread.py           :58   | test_total_space_limit_of_commitlog_with_medium_limit: Error attempting to reconnect to 127.0.57.1:9042, scheduling retry in 8.88 seconds: [Errno 111] Tried connecting to [('127.0.57.1', 9042)]. Last error: Connection refused
03:06:27,038 390     cassandra.pool                 WARNING  thread.py           :58   | test_total_space_limit_of_commitlog_with_medium_limit: Error attempting to reconnect to 127.0.57.1:9042, scheduling retry in 16.96 seconds: [Errno 111] Tried connecting to [('127.0.57.1', 9042)]. Last error: Connection refused
03:06:44,065 390     cassandra.pool                 WARNING  thread.py           :58   | test_total_space_limit_of_commitlog_with_medium_limit: Error attempting to reconnect to 127.0.57.1:9042, scheduling retry in 28.48 seconds: [Errno 111] Tried connecting to [('127.0.57.1', 9042)]. Last error: Connection refused
03:07:12,610 390     cassandra.pool                 WARNING  thread.py           :58   | test_total_space_limit_of_commitlog_with_medium_limit: Error attempting to reconnect to 127.0.57.1:9042, scheduling retry in 58.24 seconds: [Errno 111] Tried connecting to [('127.0.57.1', 9042)]. Last error: Connection refused
03:08:10,903 390     cassandra.pool                 WARNING  thread.py           :58   | test_total_space_limit_of_commitlog_with_medium_limit: Error attempting to reconnect to 127.0.57.1:9042, scheduling retry in 121.6 seconds: [Errno 111] Tried connecting to [('127.0.57.1', 9042)]. Last error: Connection refused
03:10:12,575 390     cassandra.pool                 WARNING  thread.py           :58   | test_total_space_limit_of_commitlog_with_medium_limit: Error attempting to reconnect to 127.0.57.1:9042, scheduling retry in 294.4 seconds: [Errno 111] Tried connecting to [('127.0.57.1', 9042)]. Last error: Connection refused
03:15:02,338 390     errors                         ERROR    conftest.py         :199  | test_total_space_limit_of_commitlog_with_medium_limit: test failed: 
self = <ccmlib.scylla_node.ScyllaNode object at 0x7fb4e6f4d490>
args = ['/jenkins/workspace/scylla-master/dtest-daily-release/scylla/.dtest/dtest-kulhml98/test/node1/bin/scylla', '--options...-daily-release/scylla/.dtest/dtest-kulhml98/test/node1/conf/scylla.yaml', '--log-to-stdout', '1', '--api-address', ...]
marks = [], update_pid = True, wait_other_notice = True
wait_for_binary_proto = True
ext_env = {'ASAN_OPTIONS': 'disable_coredump=0:abort_on_error=1', 'BOOST_TEST_CATCH_SYSTEM_ERRORS': 'no', 'UBSAN_OPTIONS': 'halt_on_error=1:abort_on_error=1'}
timeout = None

    def _start_scylla(self, args, marks, update_pid, wait_other_notice,
                      wait_for_binary_proto, ext_env, timeout=None):
        log_file = os.path.join(self.get_path(), 'logs', 'system.log')
        # In case we are restarting a node
        # we risk reading the old cassandra.pid file
        self._delete_old_pid()
    
        try:
            env_copy = self._launch_env
        except AttributeError:
            env_copy = os.environ
        env_copy['SCYLLA_HOME'] = self.get_path()
        env_copy.update(ext_env)
    
        with open(log_file, 'a') as scylla_log:
            self._process_scylla = \
                subprocess.Popen(args, stdout=scylla_log, stderr=scylla_log, close_fds=True, env=env_copy)
        self._process_scylla.poll()
        # When running on ccm standalone, the waiter thread would block
        # the create commands. Besides in that mode, waiting is unnecessary,
        # since the original popen reference is garbage collected.
        standalone = os.environ.get('SCYLLA_CCM_STANDALONE', None)
        if standalone is None:
            self._process_scylla_waiter = threading.Thread(target=self._wait_for_scylla)
            # Don't block the main thread on abnormal shutdown
            self._process_scylla_waiter.daemon = True
            self._process_scylla_waiter.start()
        pid_filename = os.path.join(self.get_path(), 'cassandra.pid')
        with open(pid_filename, 'w') as pid_file:
            pid_file.write(str(self._process_scylla.pid))
    
        if update_pid:
            self._update_pid(self._process_scylla)
            if not self.is_running():
                raise NodeError("Error starting node %s" % self.name,
                                self._process_scylla)
    
        if wait_other_notice:
            for node, mark in marks:
                t = timeout if timeout is not None else 120 if self.cluster.scylla_mode != 'debug' else 360
                node.watch_log_for_alive(self, from_mark=mark, timeout=t)
    
        if wait_for_binary_proto:
            try:
                t = timeout * 4 if timeout is not None else 420 if self.cluster.scylla_mode != 'debug' else 900
>               self.wait_for_binary_interface(from_mark=self.mark, process=self._process_scylla, timeout=t)

../scylla/.local/lib/python3.9/site-packages/ccmlib/scylla_node.py:296: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <ccmlib.scylla_node.ScyllaNode object at 0x7fb4e6f4d490>
kwargs = {'from_mark': 123810, 'process': <Popen: returncode: None args: ['/jenkins/workspace/scylla-master/dtest-dail...>, 'timeout': 420}

    def wait_for_binary_interface(self, **kwargs):
        """
        Waits for the Binary CQL interface to be listening.  If > 1.2 will check
        log for 'Starting listening for CQL clients' before checking for the
        interface to be listening.
    
        Emits a warning if not listening after 10 seconds.
        """
        if parse_version(self.cluster.version()) >= parse_version('1.2'):
>           self.watch_log_for("Starting listening for CQL clients", **kwargs)

../scylla/.local/lib/python3.9/site-packages/ccmlib/node.py:532: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <ccmlib.scylla_node.ScyllaNode object at 0x7fb4e6f4d490>
exprs = 'Starting listening for CQL clients', from_mark = 123810, timeout = 420
process = <Popen: returncode: None args: ['/jenkins/workspace/scylla-master/dtest-dail...>
verbose = False, filename = 'system.log'

    def watch_log_for(self, exprs, from_mark=None, timeout=600, process=None, verbose=False, filename='system.log'):
        """
        Watch the log until one or more (regular) expression are found.
        This methods when all the expressions have been found or the method
        timeouts (a TimeoutError is then raised). On successful completion,
        a list of pair (line matched, match object) is returned.
        """
        elapsed = 0
        tofind = [exprs] if isinstance(exprs, string_types) else exprs
        tofind = [re.compile(e) for e in tofind]
        matchings = []
        reads = ""
        if len(tofind) == 0:
            return None
    
        log_file = os.path.join(self.get_path(), 'logs', filename)
        while not os.path.exists(log_file):
            time.sleep(.1)
            if process:
                process.poll()
                if process.returncode is not None:
                    self.print_process_output(self.name, process, verbose)
                    if process.returncode != 0:
                        raise RuntimeError()  # Shouldn't reuse RuntimeError but I'm lazy
    
        with open(log_file) as f:
            if from_mark:
                f.seek(from_mark)
    
            while True:
                # First, if we have a process to check, then check it.
                # Skip on Windows - stdout/stderr is cassandra.bat
                if not common.is_win():
                    if process:
                        process.poll()
                        if process.returncode is not None:
                            self.print_process_output(self.name, process, verbose)
                            if process.returncode != 0:
                                raise RuntimeError()  # Shouldn't reuse RuntimeError but I'm lazy
    
                line = f.readline()
                if line:
                    reads = reads + line
                    for e in tofind:
                        m = e.search(line)
                        if m:
                            matchings.append((line, m))
    
                            # This is bogus! - One should not modify the list from inside the loop which iterates on
                            # that list.
                            # However since we are going to break from the loop a few lines below that should be ok.
                            tofind.remove(e)
    
                            if len(tofind) == 0:
                                return matchings[0] if isinstance(exprs, string_types) else matchings
                            break
                else:
                    # yep, it's ugly
                    time.sleep(0.01)
                    elapsed = elapsed + 1
                    if elapsed > 100 * timeout:
>                       raise TimeoutError(time.strftime("%d %b %Y %H:%M:%S", time.gmtime()) + " [" + self.name + "] Missing: " + str(
                            [e.pattern for e in tofind]) + ":\n" + reads[:50] + ".....\nSee {} for remainder".format(filename))
E                       ccmlib.node.TimeoutError: 06 Aug 2022 03:13:02 [node1] Missing: ['Starting listening for CQL clients']:
E                       Scylla version 5.2.0~dev-0.20220805.fbbe2529c11d w.....
E                       See system.log for remainder

../scylla/.local/lib/python3.9/site-packages/ccmlib/node.py:485: TimeoutError

During handling of the above exception, another exception occurred:

self = <commitlog_test.TestCommitLog object at 0x7fb4e6f4dc70>

    def test_total_space_limit_of_commitlog_with_medium_limit(self):
        """
        Test with 100M commitlog files, total space limit is 1024M
        """
>       self.test_total_space_limit_of_commitlog(commitlog_segment_size_in_mb=100,
                                                 commitlog_total_space_in_mb=1024)

commitlog_test.py:954: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
commitlog_test.py:935: in test_total_space_limit_of_commitlog
    node1.start(wait_for_binary_proto=True)
../scylla/.local/lib/python3.9/site-packages/ccmlib/scylla_node.py:605: in start
    scylla_process = self._start_scylla(args, marks, update_pid,
../scylla/.local/lib/python3.9/site-packages/ccmlib/scylla_node.py:298: in _start_scylla
    if not self.wait_for_starting(from_mark=self.mark):
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <ccmlib.scylla_node.ScyllaNode object at 0x7fb4e6f4d490>
from_mark = 123810, timeout = 120

    def wait_for_starting(self, from_mark=None, timeout=120):
        if from_mark is None:
            from_mark = self.mark_log()
        process=self._process_scylla
        starting_message = 'Starting listening for CQL clients'
        bootstrap_message = r'storage_service .* Starting to bootstrap'
        resharding_message = r'(compaction|database) -.*Resharding'
        if not self.watch_log_for("{}|{}|{}".format(starting_message, bootstrap_message, resharding_message), from_mark=from_mark, timeout=timeout, process=process):
            return False
        prev_mark = from_mark
        prev_mark_time = time.time()
        sleep_time = 10 if timeout >= 100 else 1
        while not self.grep_log(starting_message, from_mark=from_mark):
            process.poll()
            if process.returncode is not None:
                self.print_process_output(self.name, process, verbose=True)
                if process.returncode != 0:
                    raise RuntimeError("The process is dead, returncode={}".format(process.returncode))
            repair_pattern = r'repair - Repair \d+ out of \d+ ranges'
            streaming_pattern = r'range_streamer - Bootstrap .* streaming .* ranges'
            resharding_pattern = r'(compaction|database) -.*Resharded'
            if self.grep_log("{}|{}|{}".format(repair_pattern, streaming_pattern, resharding_pattern), from_mark=prev_mark):
                prev_mark = self.mark_log()
                prev_mark_time = time.time()
            elif time.time() - prev_mark_time >= timeout:
>               raise TimeoutError("{}: Timed out waiting for '{}'".format(self.name, starting_message))
E               ccmlib.node.TimeoutError: node1: Timed out waiting for 'Starting listening for CQL clients'

../scylla/.local/lib/python3.9/site-packages/ccmlib/scylla_node.py:247: TimeoutError
03:15:02,338 390     errors                         ERROR    conftest.py         :199  | test_total_space_limit_of_commitlog_with_medium_limit: test failed: 
self = <ccmlib.scylla_node.ScyllaNode object at 0x7fb4e6f4d490>
args = ['/jenkins/workspace/scylla-master/dtest-daily-release/scylla/.dtest/dtest-kulhml98/test/node1/bin/scylla', '--options...-daily-release/scylla/.dtest/dtest-kulhml98/test/node1/conf/scylla.yaml', '--log-to-stdout', '1', '--api-address', ...]
marks = [], update_pid = True, wait_other_notice = True
wait_for_binary_proto = True
ext_env = {'ASAN_OPTIONS': 'disable_coredump=0:abort_on_error=1', 'BOOST_TEST_CATCH_SYSTEM_ERRORS': 'no', 'UBSAN_OPTIONS': 'halt_on_error=1:abort_on_error=1'}
timeout = None

    def _start_scylla(self, args, marks, update_pid, wait_other_notice,
                      wait_for_binary_proto, ext_env, timeout=None):
        log_file = os.path.join(self.get_path(), 'logs', 'system.log')
        # In case we are restarting a node
        # we risk reading the old cassandra.pid file
        self._delete_old_pid()
    
        try:
            env_copy = self._launch_env
        except AttributeError:
            env_copy = os.environ
        env_copy['SCYLLA_HOME'] = self.get_path()
        env_copy.update(ext_env)
    
        with open(log_file, 'a') as scylla_log:
            self._process_scylla = \
                subprocess.Popen(args, stdout=scylla_log, stderr=scylla_log, close_fds=True, env=env_copy)
        self._process_scylla.poll()
        # When running on ccm standalone, the waiter thread would block
        # the create commands. Besides in that mode, waiting is unnecessary,
        # since the original popen reference is garbage collected.
        standalone = os.environ.get('SCYLLA_CCM_STANDALONE', None)
        if standalone is None:
            self._process_scylla_waiter = threading.Thread(target=self._wait_for_scylla)
            # Don't block the main thread on abnormal shutdown
            self._process_scylla_waiter.daemon = True
            self._process_scylla_waiter.start()
        pid_filename = os.path.join(self.get_path(), 'cassandra.pid')
        with open(pid_filename, 'w') as pid_file:
            pid_file.write(str(self._process_scylla.pid))
    
        if update_pid:
            self._update_pid(self._process_scylla)
            if not self.is_running():
                raise NodeError("Error starting node %s" % self.name,
                                self._process_scylla)
    
        if wait_other_notice:
            for node, mark in marks:
                t = timeout if timeout is not None else 120 if self.cluster.scylla_mode != 'debug' else 360
                node.watch_log_for_alive(self, from_mark=mark, timeout=t)
    
        if wait_for_binary_proto:
            try:
                t = timeout * 4 if timeout is not None else 420 if self.cluster.scylla_mode != 'debug' else 900
>               self.wait_for_binary_interface(from_mark=self.mark, process=self._process_scylla, timeout=t)

../scylla/.local/lib/python3.9/site-packages/ccmlib/scylla_node.py:296: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <ccmlib.scylla_node.ScyllaNode object at 0x7fb4e6f4d490>
kwargs = {'from_mark': 123810, 'process': <Popen: returncode: None args: ['/jenkins/workspace/scylla-master/dtest-dail...>, 'timeout': 420}

    def wait_for_binary_interface(self, **kwargs):
        """
        Waits for the Binary CQL interface to be listening.  If > 1.2 will check
        log for 'Starting listening for CQL clients' before checking for the
        interface to be listening.
    
        Emits a warning if not listening after 10 seconds.
        """
        if parse_version(self.cluster.version()) >= parse_version('1.2'):
>           self.watch_log_for("Starting listening for CQL clients", **kwargs)

../scylla/.local/lib/python3.9/site-packages/ccmlib/node.py:532: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <ccmlib.scylla_node.ScyllaNode object at 0x7fb4e6f4d490>
exprs = 'Starting listening for CQL clients', from_mark = 123810, timeout = 420
process = <Popen: returncode: None args: ['/jenkins/workspace/scylla-master/dtest-dail...>
verbose = False, filename = 'system.log'

    def watch_log_for(self, exprs, from_mark=None, timeout=600, process=None, verbose=False, filename='system.log'):
        """
        Watch the log until one or more (regular) expression are found.
        This methods when all the expressions have been found or the method
        timeouts (a TimeoutError is then raised). On successful completion,
        a list of pair (line matched, match object) is returned.
        """
        elapsed = 0
        tofind = [exprs] if isinstance(exprs, string_types) else exprs
        tofind = [re.compile(e) for e in tofind]
        matchings = []
        reads = ""
        if len(tofind) == 0:
            return None
    
        log_file = os.path.join(self.get_path(), 'logs', filename)
        while not os.path.exists(log_file):
            time.sleep(.1)
            if process:
                process.poll()
                if process.returncode is not None:
                    self.print_process_output(self.name, process, verbose)
                    if process.returncode != 0:
                        raise RuntimeError()  # Shouldn't reuse RuntimeError but I'm lazy
    
        with open(log_file) as f:
            if from_mark:
                f.seek(from_mark)
    
            while True:
                # First, if we have a process to check, then check it.
                # Skip on Windows - stdout/stderr is cassandra.bat
                if not common.is_win():
                    if process:
                        process.poll()
                        if process.returncode is not None:
                            self.print_process_output(self.name, process, verbose)
                            if process.returncode != 0:
                                raise RuntimeError()  # Shouldn't reuse RuntimeError but I'm lazy
    
                line = f.readline()
                if line:
                    reads = reads + line
                    for e in tofind:
                        m = e.search(line)
                        if m:
                            matchings.append((line, m))
    
                            # This is bogus! - One should not modify the list from inside the loop which iterates on
                            # that list.
                            # However since we are going to break from the loop a few lines below that should be ok.
                            tofind.remove(e)
    
                            if len(tofind) == 0:
                                return matchings[0] if isinstance(exprs, string_types) else matchings
                            break
                else:
                    # yep, it's ugly
                    time.sleep(0.01)
                    elapsed = elapsed + 1
                    if elapsed > 100 * timeout:
>                       raise TimeoutError(time.strftime("%d %b %Y %H:%M:%S", time.gmtime()) + " [" + self.name + "] Missing: " + str(
                            [e.pattern for e in tofind]) + ":\n" + reads[:50] + ".....\nSee {} for remainder".format(filename))
E                       ccmlib.node.TimeoutError: 06 Aug 2022 03:13:02 [node1] Missing: ['Starting listening for CQL clients']:
E                       Scylla version 5.2.0~dev-0.20220805.fbbe2529c11d w.....
E                       See system.log for remainder

../scylla/.local/lib/python3.9/site-packages/ccmlib/node.py:485: TimeoutError

During handling of the above exception, another exception occurred:

self = <commitlog_test.TestCommitLog object at 0x7fb4e6f4dc70>

    def test_total_space_limit_of_commitlog_with_medium_limit(self):
        """
        Test with 100M commitlog files, total space limit is 1024M
        """
>       self.test_total_space_limit_of_commitlog(commitlog_segment_size_in_mb=100,
                                                 commitlog_total_space_in_mb=1024)

commitlog_test.py:954: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
commitlog_test.py:935: in test_total_space_limit_of_commitlog
    node1.start(wait_for_binary_proto=True)
../scylla/.local/lib/python3.9/site-packages/ccmlib/scylla_node.py:605: in start
    scylla_process = self._start_scylla(args, marks, update_pid,
../scylla/.local/lib/python3.9/site-packages/ccmlib/scylla_node.py:298: in _start_scylla
    if not self.wait_for_starting(from_mark=self.mark):
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <ccmlib.scylla_node.ScyllaNode object at 0x7fb4e6f4d490>
from_mark = 123810, timeout = 120

    def wait_for_starting(self, from_mark=None, timeout=120):
        if from_mark is None:
            from_mark = self.mark_log()
        process=self._process_scylla
        starting_message = 'Starting listening for CQL clients'
        bootstrap_message = r'storage_service .* Starting to bootstrap'
        resharding_message = r'(compaction|database) -.*Resharding'
        if not self.watch_log_for("{}|{}|{}".format(starting_message, bootstrap_message, resharding_message), from_mark=from_mark, timeout=timeout, process=process):
            return False
        prev_mark = from_mark
        prev_mark_time = time.time()
        sleep_time = 10 if timeout >= 100 else 1
        while not self.grep_log(starting_message, from_mark=from_mark):
            process.poll()
            if process.returncode is not None:
                self.print_process_output(self.name, process, verbose=True)
                if process.returncode != 0:
                    raise RuntimeError("The process is dead, returncode={}".format(process.returncode))
            repair_pattern = r'repair - Repair \d+ out of \d+ ranges'
            streaming_pattern = r'range_streamer - Bootstrap .* streaming .* ranges'
            resharding_pattern = r'(compaction|database) -.*Resharded'
            if self.grep_log("{}|{}|{}".format(repair_pattern, streaming_pattern, resharding_pattern), from_mark=prev_mark):
                prev_mark = self.mark_log()
                prev_mark_time = time.time()
            elif time.time() - prev_mark_time >= timeout:
>               raise TimeoutError("{}: Timed out waiting for '{}'".format(self.name, starting_message))
E               ccmlib.node.TimeoutError: node1: Timed out waiting for 'Starting listening for CQL clients'

../scylla/.local/lib/python3.9/site-packages/ccmlib/scylla_node.py:247: TimeoutError
03:17:09,545 390     ccm                            WARNING  cluster.py          :710  | test_total_space_limit_of_commitlog_with_medium_limit: node1: node1 is still running after 127 seconds. Trying to generate coredump using kill(32900, SIGQUIT)...
03:17:10,547 390     ccm                            WARNING  cluster.py          :710  | test_total_space_limit_of_commitlog_with_medium_limit: node1: node1 scylla-jmx is still running. Killing process using kill(127, SIGKILL)...
03:17:10,559 390     dtest_setup                    WARNING  dtest_setup.py      :155  | test_total_space_limit_of_commitlog_with_medium_limit: Moving core file /jenkins/workspace/scylla-master/dtest-daily-release/scylla-dtest/scylla.32900.1659755829.core to /jenkins/workspace/scylla-master/dtest-daily-release/scylla-dtest/logs-full.release.001/1659755830552_commitlog_test.py::TestCommitLog::test_total_space_limit_of_commitlog_with_medium_limit/node1-scylla.32900.1659755829.core
03:17:17,638 390     dtest_setup                    DEBUG    dtest_setup.py      :609  | test_total_space_limit_of_commitlog_with_medium_limit: removing ccm cluster test at: /jenkins/workspace/scylla-master/dtest-daily-release/scylla/.dtest/dtest-kulhml98
03:17:17,840 390     dtest_setup                    DEBUG    dtest_setup.py      :613  | test_total_space_limit_of_commitlog_with_medium_limit: clearing ssl stores from [/jenkins/workspace/scylla-master/dtest-daily-release/scylla/.dtest/dtest-kulhml98] directory
03:17:17,840 390     dtest_setup                    DEBUG    dtest_setup.py      :83   | test_total_space_limit_of_commitlog_with_medium_limit: Freeing cluster ID 57: link /jenkins/workspace/scylla-master/dtest-daily-release/scylla/.dtest/57
```

--- a/db/commitlog/commitlog.cc
+++ b/db/commitlog/commitlog.cc
@@ -225,7 +225,6 @@ const std::string db::commitlog::descriptor::FILENAME_EXTENSION(".log");
 class db::commitlog::segment_manager : public ::enable_shared_from_this<segment_manager> {
 public:
     config cfg;
-    std::vector<sstring> _segments_to_replay;
     const uint64_t max_size;
     const uint64_t max_mutation_size;
     // Divide the size-on-disk threshold by #cpus used, since we assume
@@ -479,7 +478,8 @@ public:
     buffer_type acquire_buffer(size_t s, size_t align);
     temporary_buffer<char> allocate_single_buffer(size_t, size_t);
 
-    future<std::vector<descriptor>> list_descriptors(sstring dir);
+    future<std::vector<descriptor>> list_descriptors(sstring dir) const;
+    future<std::vector<sstring>> get_segments_to_replay() const;
 
     flush_handler_id add_flush_handler(flush_handler h) {
         auto id = ++_flush_ids;
@@ -499,7 +499,7 @@ private:
     void abort_recycled_list(std::exception_ptr);
 
     size_t max_request_controller_units() const;
-    segment_id_type _ids = 0;
+    segment_id_type _ids = 0, _low_id = 0;
     std::vector<sseg_ptr> _segments;
     queue<sseg_ptr> _reserve_segments;
     queue<named_file> _recycled_segments;
@@ -1436,7 +1436,7 @@ future<> db::commitlog::segment_manager::replenish_reserve() {
 }
 
 future<std::vector<db::commitlog::descriptor>>
-db::commitlog::segment_manager::list_descriptors(sstring dirname) {
+db::commitlog::segment_manager::list_descriptors(sstring dirname) const {
     auto dir = co_await open_checked_directory(commit_error_handler, dirname);
     std::vector<db::commitlog::descriptor> result;
 
@@ -1466,6 +1466,26 @@ db::commitlog::segment_manager::list_descriptors(sstring dirname) {
     co_return result;
 }
 
+// #11237 - make get_segments_to_replay on-demand. Since we base the time-part of
+// descriptor ids on hightest of wall-clock and segments found on disk on init,
+// we can just scan files now and include only those representing generations before
+// the creation of this commitlog instance.
+// This _could_ give weird results iff we had a truly sharded commitlog folder
+// where init of the instances was not very synced _and_ we allowed more than
+// one shard to do replay. But allowed usage always either is one dir - one shard (hints)
+// or does shard 0 init first (main/database), then replays on 0 as well.
+future<std::vector<sstring>> db::commitlog::segment_manager::get_segments_to_replay() const {
+    std::vector<sstring> segments_to_replay;
+    auto descs = co_await list_descriptors(cfg.commit_log_location);
+    for (auto& d : descs) {
+        auto id = replay_position(d.id).base_id();
+        if (id <= _low_id) {
+            segments_to_replay.push_back(cfg.commit_log_location + "/" + d.filename());
+        }
+    }
+    co_return segments_to_replay;
+}
+
 future<> db::commitlog::segment_manager::init() {
     auto descs = co_await list_descriptors(cfg.commit_log_location);
 
@@ -1473,16 +1493,12 @@ future<> db::commitlog::segment_manager::init() {
     segment_id_type id = *cfg.base_segment_id;
     for (auto& d : descs) {
         id = std::max(id, replay_position(d.id).base_id());
-        _segments_to_replay.push_back(cfg.commit_log_location + "/" + d.filename());
-        // #11184 - include replay footprint so we make sure to delete any segments
-        // pushing us over limits in "delete_segments" (assumed called after replay)
-        // Note: if noone calls get_segments_to_replay + delete we're rather borked,
-        // but...
-        totals.total_size_on_disk += co_await file_size(_segments_to_replay.back());
     }
 
     // base id counter is [ <shard> | <base> ]
     _ids = replay_position(this_shard_id(), id).id;
+    _low_id = id;
+
     // always run the timer now, since we need to handle segment pre-alloc etc as well.
     _timer.set_callback(std::bind(&segment_manager::on_timer, this));
     auto delay = this_shard_id() * std::ceil(double(cfg.commitlog_sync_period_in_ms) / smp::count);
@@ -2955,8 +2971,8 @@ future<std::vector<sstring>> db::commitlog::list_existing_segments(const sstring
     });
 }
 
-std::vector<sstring> db::commitlog::get_segments_to_replay() const {
-    return std::move(_segment_manager->_segments_to_replay);
+future<std::vector<sstring>> db::commitlog::get_segments_to_replay() const {
+    return _segment_manager->get_segments_to_replay();
 }
 
 future<> db::commitlog::delete_segments(std::vector<sstring> files) const {

--- a/db/commitlog/commitlog.cc
+++ b/db/commitlog/commitlog.cc
@@ -2048,9 +2048,16 @@ future<> db::commitlog::segment_manager::delete_segments(std::vector<sstring> fi
         // Note: this is only for replay files. We can decide to
         // recycle these, but they don't count into footprint,
         // thus unopened named_files are what we want (known_size == 0)
+
+        // #11184 - include replay footprint so we make sure to delete any segments
+        // pushing us over limits in "delete_segments" (assumed called after replay)
+        // #11237 - cannot do this already in "init()" - we need to be able to create
+        // segments before replay+delete, and more to the point: shards that _don't_ replay
+        // must not add this to their footprint (a.) shared, b.) not there after this call)
+        totals.total_size_on_disk += co_await file_size(s);
         _files_to_dispose.emplace_back(s, dispose_mode::Delete);
     }
-    return do_pending_deletes();
+    co_return co_await do_pending_deletes();
 }
 
 void db::commitlog::segment_manager::abort_recycled_list(std::exception_ptr ep) {

--- a/db/commitlog/commitlog.hh
+++ b/db/commitlog/commitlog.hh
@@ -270,10 +270,8 @@ public:
     /**
      * Returns a vector of segment paths which were
      * preexisting when this instance of commitlog was created.
-     *
-     * The list will be empty when called for the second time.
      */
-    std::vector<sstring> get_segments_to_replay() const;
+    future<std::vector<sstring>> get_segments_to_replay() const;
 
     /**
      * Delete aforementioned segments, and possible metadata

--- a/db/hints/manager.cc
+++ b/db/hints/manager.cc
@@ -425,14 +425,14 @@ future<db::commitlog> manager::end_point_hints_manager::add_store() noexcept {
             // which is larger than the segment ID of the RP of the last written hint.
             cfg.base_segment_id = _last_written_rp.base_id();
 
-            return commitlog::create_commitlog(std::move(cfg)).then([this] (commitlog l) {
+            return commitlog::create_commitlog(std::move(cfg)).then([this] (commitlog l) -> future<commitlog> {
                 // add_store() is triggered every time hint files are forcefully flushed to I/O (every hints_flush_period).
                 // When this happens we want to refill _sender's segments only if it has finished with the segments he had before.
                 if (_sender.have_segments()) {
-                    return make_ready_future<commitlog>(std::move(l));
+                    co_return l;
                 }
 
-                std::vector<sstring> segs_vec = l.get_segments_to_replay();
+                std::vector<sstring> segs_vec = co_await l.get_segments_to_replay();
 
                 if (segs_vec.empty()) {
                     // If the segs_vec is empty, this means that there are no more
@@ -448,7 +448,7 @@ future<db::commitlog> manager::end_point_hints_manager::add_store() noexcept {
                     auto rp = _last_written_rp;
                     rp.pos++;
                     _sender.rewind_sent_replay_position_to(rp);
-                    return make_ready_future<commitlog>(std::move(l));
+                    co_return l;
                 }
 
                 std::vector<std::pair<db::segment_id_type, sstring>> local_segs_vec;
@@ -474,7 +474,7 @@ future<db::commitlog> manager::end_point_hints_manager::add_store() noexcept {
                     _sender.add_segment(std::move(seg));
                 }
 
-                return make_ready_future<commitlog>(std::move(l));
+                co_return l;
             });
         });
     });

--- a/main.cc
+++ b/main.cc
@@ -1138,7 +1138,7 @@ To start the scylla server proper, simply invoke as: scylla server (or just scyl
 
             auto sch_cl = db.local().schema_commitlog();
             if (sch_cl != nullptr) {
-                auto paths = sch_cl->get_segments_to_replay();
+                auto paths = sch_cl->get_segments_to_replay().get();
                 if (!paths.empty()) {
                     supervisor::notify("replaying schema commit log");
                     auto rp = db::commitlog_replayer::create_replayer(db).get0();
@@ -1164,7 +1164,7 @@ To start the scylla server proper, simply invoke as: scylla server (or just scyl
             supervisor::notify("starting commit log");
             auto cl = db.local().commitlog();
             if (cl != nullptr) {
-                auto paths = cl->get_segments_to_replay();
+                auto paths = cl->get_segments_to_replay().get();
                 if (!paths.empty()) {
                     supervisor::notify("replaying commit log");
                     auto rp = db::commitlog_replayer::create_replayer(db).get0();

